### ISSUE TITLE
[AH-28] Enhance application to follow A11y criterias

### ISSUE THREAD

**Issue by Shagon1k** [Shagon1k/AHurynovich-CV#49]
### Description
!! Work as **_AHurynovich-Accessibility or simply_ [AH-A]**

**Important** to start with not "how" and "what" but with "why" and "for whom": check different disabilities types, how users struggle with them, which tools do they use, etc.

https://developer.mozilla.org/en-US/docs/Web/CSS/@media/prefers-reduced-motion

!! **Remember!** 
The important thing to understand about the TAB key, is that it is used to move from one operable item to the next. Any non-operable content in between will be ignored. So if a paragraph has some links inside a sentence, pressing TAB will just read out the links, not the rest of the sentence. The TAB key also stops at buttons and form inputs.
When the user wants to read the whole paragraph, they will proceed using the arrow keys ("browse" mode of Screen Readers, not "focus" mode). When dealing with a form or application-like UI, they can use the TAB key. 
Page elements that are not interactive to mouse or touch users should not be made keyboard focusable (such as by using tabindex).
[More details](https://ux.stackexchange.com/questions/140358/should-screen-readers-read-the-content-without-tab-or-with-tab)
[More details 2](https://webaim.org/techniques/keyboard/)

! **Do not use tabindex values of 1 or greater** to change the default keyboard navigation order. Think about logical TAB order (top-to-bottom, left-to-right) from the beginning, then left the Browser do all the job. Tab indexation change should be done **only for exceptional cases**.

#### W3C course with certification
https://www.w3.org/WAI/fundamentals/foundations-course/
https://courses.edx.org/certificates/ff5e044720d34b75922eecaf01440906

#### Guidelines
![image](https://user-images.githubusercontent.com/11179843/142912146-bb65ecc1-917c-4792-84cd-5e52ff518d1c.png)

**_Perceivable - ability to find content:_**
<img src="https://user-images.githubusercontent.com/11179843/141789850-c6032432-e079-4315-a762-c49f547e33b4.png" width="100px"/>
 - **Text alternatives** (all non-text content has it's alternatives, **not only images** but charts, diagrams, buttons, etc.) Note: Should be described not only based on content inside, but also content which it covers should be kept in mind + image functionality e.g. "My page homepage logo". Some images can have "alt" skipped in case adding it will only increase complexity (e.g. divider image) Remember: skipping alt !== having null one. Alt not null: informative, actionable images, alt null: decorative images.
 - **Media alternatives** (videos\audios has captions, subtitles, text transcripts, audio descriptors, sign language)
 - **Adaptable** (acceptable for different devices and it's orientations without lost information and structure (context) + well-read by Screen Readers (good Semantic, e.g. do not skip headings) )
 - **Distinguishable** (content (text, images) can be separated from it's background - use https://contrast-ratio.com/ OR https://webaim.org/resources/contrastchecker/ OR https://color.adobe.com/create/color-accessibility . Recommended ration is 3-4.5 (based on large/small content))

 **_Operable - ability to use application:_**
<img src="https://user-images.githubusercontent.com/11179843/141789876-74982345-7208-4fbf-9497-d1c2ef80d9b4.png" width="100px"/>
 - **Keyboard accessible** (all parts of app accessible with only a keyboard. Remember: buttons for _actions_, links for _moving to some content._ AVOID keyboard traps. Focus locker or similar package could be used for modals\popups\navigation case.) TAB RULE: tab all, tab away, tab order, visual focus, all functionality by focus;
 - **Enough time** (all content should have enough time to be watched)
 - **Avoid seizures and physical reactions** (do not design content in a way that is known to cause seizures or physical reactions) So avoid flash more than 3 times in a second, animation covering large area of the screen, bright ones.
 - **Navigable** (easy to navigate and understand - more about semantic) Note: unique <title> for each page is good practice (dash "|" can be used for complex titles, precise title first: "Contact Us | My Company") -> React Helmet
 - **Input Modalities** (all parts of app accessible with mouse\touch)

**_Understandable - ability to understand each part of app (semantic):_**
<img src="https://user-images.githubusercontent.com/11179843/141789935-e9efe288-989a-4454-852f-296be655e418.png" width="100px"/>
 - **Readable** (content is readable and easy to understand. Abbreviations could use <abbr> tag )
 - **Predictable** (app is predictable in how it operate. E.g. base structure is kept over the page and app navigation)
 - **Input assistanc**e (easy to correct and avoid mistakes - validation and tips)

 **_Robust - be compatible:_**
<img src="https://user-images.githubusercontent.com/11179843/141789962-947236a3-fc23-4e29-ab45-4e8406b16eda.png" width="100px"/>
 - **Compatible** (working for current and future accesstive technology):
   Write valid, semantic HTML (e.g. H1-H5 matters, HTML tags matters, etc.)
   Use ARIA labels to extend HTML (more applicable for interactive components)
   Use a11y testing tools

**More information [here](https://www.w3.org/WAI/WCAG21/quickref/#principle1)**

### Notes
**Tips:**
 - Once A11y plan created, it is worth to add a11y statement and place it somewhere on your web site (e.g. in Footer) in order to notify users that you follow it. Statement generator link: https://www.w3.org/WAI/planning/statements/generator/#create (additional [info link](https://learning.edx.org/course/course-v1:W3Cx+WAI0.1x+3T2019/block-v1:W3Cx+WAI0.1x+3T2019+type@sequential+block@438c08dbef4b4230a1d7c941b7396cde/block-v1:W3Cx+WAI0.1x+3T2019+type@vertical+block@990f4ff2bcda494892da992863c4722e))
 - Add a11y to **CI\CD pipeline**
 - Aria labels to check and add (aria-controls, aria-expanded, aria-role etc.) 

<details>
 <summary>Details</summary>
<img src="https://user-images.githubusercontent.com/11179843/141797271-a4242284-0cac-4c89-a43d-0d41e25954fa.png"/>
</details>

 - Identify inputs and help browser autocomplete

<details>
 <summary>Details</summary>
<img src="https://user-images.githubusercontent.com/11179843/141794641-bf44f85c-cfbf-44af-a3a0-44bf17ff0911.png"/>
</details> 

 - Zooming (at least 200-300%) should be taken in to a count. **Note:** having 2 scrolling (vertical + horizontal) are not welcomed as in such way page become not so accessible and harder to read and understand
 - Lower zoom (e.g. 50%) also matters for the people with tunnel vision
 - Orientation should be taken in to a count
 - Think about ability to remove animation (@no-animation mixin could be created)
 
<details>
 <summary>Details</summary>
<img src="https://user-images.githubusercontent.com/11179843/141798285-586ccaa5-e220-4103-b098-a605aeebe394.png"/>
</details>

 - New AAA standard is to buttons be at least 44x44
 - For ON/OFF switchers or something like examples text notations should be used

<details>
 <summary>Details</summary>
<img src="https://user-images.githubusercontent.com/11179843/141863061-39378f6c-373b-4e25-b55d-ff72cce5fb28.png"/>
</details> 

 - Use "sr-only" elements for explanations. Such elements should also keep semantic (Headers, spans, navs, etc.)

<details>
 <summary>Details</summary>
<img src="https://user-images.githubusercontent.com/11179843/141864593-4f98a084-5c0c-4c3f-8723-f0077a37573d.png"/>
<img src="https://user-images.githubusercontent.com/11179843/141865002-67c43b76-a0b4-48c8-bc55-8372b2702f3e.png"/>
(Simple unseen elements could not be used as browser identifies them as "not needed" and possibly just skip)
<img src="https://user-images.githubusercontent.com/11179843/141867045-405f35aa-e3ad-41b6-8789-77e19cc3cc28.png"/>

</details> 

 - Use "aria-hidden=true" ARIA label in order to Screen Reader to skip content (in case no need to pass)
 - Accessible SVG
<details>
 <summary>Details</summary>
<img src="https://user-images.githubusercontent.com/11179843/141974904-4d2bb975-e728-4a07-a5e8-ea5364baa73b.png"/>
</details> 

 - Video embeds should not load on page load, only on demand need. Iframe with "srcdoc" attribute settled up could be used to handle this. DO NOT FORGET ABOUT TRANSCRIPTS FOR A11Y
 
<details>
 <summary>Details</summary>
<img src="https://user-images.githubusercontent.com/11179843/141975651-71564e13-6696-4ff8-b193-66559df999f7.png"/>
</details> 

 - **FireFox browser has much better A11y built-in tools**, **Web Developer** Chrome could be used as alternative
 - **Wave** extension is really good one
 - Additionally **Chrome Lighthouse** could be used to test A11y
 - **Screen Reader extension**
 - Sometimes it is worth to add "Skip smth" hidden link (e.g. "Skip navigation")
 
<details>
 <summary>Details</summary>
<img src="https://user-images.githubusercontent.com/11179843/142014893-ac5bde5e-fcce-4d4d-9c50-65083aa0fbf0.png"/>
</details> 

 - Managing programmatically `onBlur` event it is worth to use `setTimeout` as wrapper to callback. It is valuable as we need to first check if another child of the element has received focus as the blur event fires prior to the new focus event
 - Also worth to add some **a11y linters** like [jsx-eslint/jsx-a11y](https://github.com/jsx-eslint/eslint-plugin-jsx-a11y), [stylelint-a11y](https://github.com/YozhikM/stylelint-a11y)
 - Don't forget to use `"role"` attribute in addition to ARIA labels

<details>
 <summary>Details</summary>
<img src="https://user-images.githubusercontent.com/11179843/142042116-b405406b-a4e9-4d41-9206-c2a9c9bb4a2c.png"/>
</details> 

 - Test a11y also using [testing-library/dom](https://github.com/testing-library/dom-testing-library), [testing-library/jest-dom](https://testing-library.com/docs/ecosystem-jest-dom/), [user-event](https://testing-library.com/docs/ecosystem-user-event/)
Inspire example: https://github.com/denis-biruk/a11y-test-tabs

<details>
 <summary>Details</summary>
<img src="https://user-images.githubusercontent.com/11179843/142043577-8309d45d-07ac-49fc-b6bc-9c6c547f4368.png"/>
<img src="https://user-images.githubusercontent.com/11179843/142043671-1ad24455-84eb-4fdc-beda-ca57dc024b98.png"/>
<img src="https://user-images.githubusercontent.com/11179843/142043947-67003ea9-20cb-4bc7-9bc3-33f753fec260.png"/>
</details> 

- Check google.com tabulation for inspiration about skip link and "how to" for people with disabilities

 - Tooltips should not contain really important information as it is not always accessible (e.g. for users using Screen Magnification)
 - Use a11y anchors to skip over the site for more welcomed interface:
![image](https://user-images.githubusercontent.com/11179843/142505290-860bea1b-957c-4f2b-975a-509171d674c7.png)
 - Android TalkBack feature can be also used to check site A11y
- When **opening Popup** the focus should automatically go to "X" button (close) to provide USER easy possibility to close in case unintentional opening (the **only exception** happens if there are some focusable items before "X" button, then focus should be set to them in order not to skip when TAB navigate);
- **Button** for the ***action***, **Link** for the **navigation**;
- NO necessity to specify both "ariaLabel" and "title": in such case both will be read by Screen Reader;
- avoid `outline: 0`\`outline: none`. In addition to the default outline, you can use CSS to make the focus indicator more visually apparent and keyboard-friendly by ensuring the focus indicator is highly visible with sufficient contrast, and by adding a background color or other visual focus style to links and other interactive controls. The outline can be styled to match your site design but should be readily detected when navigating with the Tab key.

#### A11y settings popup
- A11y settings popup could be created as an option. It could be used e.g. for changing text font to more accessible one, remove animations, activate more color-friendly theme, change font size, change basic margins, text line-height and line-spacing, reduced number of images and content etc.
- CSS-specific variables (e.g. preferable font size) could be stored by A11y settings service in CSS variables to automatically work only on CSS basis (https://developer.mozilla.org/en-US/docs/Web/CSS/Using_CSS_custom_properties)
- Opening popup will display a list of different disabilities people might have (cognitive, low visual, absence of visual, etc.) and application will apply changes based on chosen

#### A11y CI/CD automation
There are a number of tools. Worth to check and add to pipeline. **Example:** [axe-core](https://github.com/dequelabs/axe-core)
Cypress also could be used to test A11y in a couple with Axe. https://www.npmjs.com/package/cypress-axe
